libs/platform/kernel/registry_helper_km.c

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "ebpf_registry_helper.h"
+
+void
+close_registry_key(ebpf_registry_key_t key)
+{
+    ZwClose(key);
+}
+
+_Success_(return == 0) NTSTATUS
+    convert_guid_to_string(_In_ const GUID* guid, _Out_writes_all_(string_size) wchar_t* string, size_t string_size)
+{
+    UNICODE_STRING unicode_string;
+    NTSTATUS status = RtlStringFromGUID(guid, &unicode_string);
+    if (status != STATUS_SUCCESS) {
+        goto Exit;
+    }
+
+    if (string_size < GUID_STRING_LENGTH + 1) {
+        status = STATUS_BUFFER_TOO_SMALL;
```

Comment:
is ```string_size``` specified in multiples of bytes or ```wchar_t```? If former, shouldn't this check be in terms of ```wchar_t``` multiples, i.e., ```if (string_size < ((GUID_STRING_LENGTH * sizeof(wchar_t)) + sizeof(wchar_t)))```? we are copying _bytes_ in the ```memcpy``` call below and at first blush I think we run the risk of he buffer overrun.

Am I missing something here?

Reply:
I filed issue https://github.com/microsoft/ebpf-for-windows/issues/2123, which captures addressing this as well as other hardening specific to the existing registry APIs.

Reply:
I'd disagree here as well :-) I wouldn't classify this as 'hardening' as this is really a bug in the _new_ code introduced by this PR.  _My_ vote re. code feedback with _new_ code is it ought to be addressed prior to merging.  This has pretty much been the standard practice that I've been following (and have seen others follow) in the eBPF-for-windows project.

Yes, there _have_ been instances where late breaking problems were addressed under a new issue, but those were almost always for last-minute design and/or implementation edge cases that required non-trivial code changes, but never for code bugs.  Adding potential bugs to that bucket IMO is a very slippery slope.

Reply:
> I'd disagree here as well :-) I wouldn't classify this as 'hardening' as this is really a bug in the _new_ code introduced by this PR. 

Just to make sure it's actually clear: this PR does *not* introduce new code. 

Specifically, the PR simply *moves* the existing implementation within a header file (which is not ideal btw), i.e. `\include\kernel\ebpf_registry_helper.h` into a separate c file, i.e. `libs/platform/kernel/registry_helper_km.c`.

That said, the idea of separating the feedback to scope the PRs we release, is to have specific (and eventually easily reversible) PRs to the source issues they are addressing, while *absolutely tracking all feedback in this PR*.

Moreover, *the new issue will apply this feedback to even other source code* that is not involved in this PR as well.

Happy to go by the team rules here.

Reply:
`Specifically, the PR simply moves the existing implementation within a header file (which is not ideal btw), i.e. \include\kernel\ebpf_registry_helper.h into a separate c file, i.e. libs/platform/kernel/registry_helper_km.c.`

I think moving to `libs/platform/kernel/registry_helper_km.c` is wrong.
1. `\include\ebpf_store_helper.h` and `\include\kernel\registry_helper.h` need to be part of nuget package so that anyone writing a new extension can include the header files and use the store APIs to update the store. Either we keep the function implementations in `ebpf_registry_helper.h` and include the header file in the nuget package or create a new ebpf_store.lib and include that in the nuget package (along with the header file). 
2. Even though the functions implemented in `\include\ebpf_store_helper.h` are marked static (because they are implemented in a header file, and will give build errors if the same header file is included in multiple C files), they should not start with `_`.

Reply:
> I think moving to `libs/platform/kernel/registry_helper_km.c` is wrong.

I think it's not wrong.

Because of how these registry APIs were designed, splitting implementations becomes a necessity when you need to include APIs that have the same signature in both kernel mode and use mode. Otherwise, all the overlapping functions (i.e. `open_registry_key`, `delete_registry_key`, `delete_registry_tree`) would need an implementation in kernel mode that is not required nor used by any kernel binaries, unless we just want to implement them as empty/error returning values.

I do get that we want to provide those headers with implementations so that we deliver a dev kit to users, (i.e. in the MSI), but IMHO, this is not right either.

Therefore, I'm happy to revert splitting implementations from headers, and implement the added functions as empty functions returning errors - *in the header*.

What about the other points of discussion?

Reply:
> is `string_size` specified in multiples of bytes or `wchar_t`? If former, shouldn't this check be in terms of `wchar_t` multiples, i.e., `if (string_size < ((GUID_STRING_LENGTH * sizeof(wchar_t)) + sizeof(wchar_t)))`? we are copying _bytes_ in the `memcpy` call below and at first blush I think we run the risk of buffer overrun.
> 
> Am I missing something here?

This ends up requiring just a replacement for hard-coded `wchar_t`'s size (i.e. `2`), and renaming the `string_size` to `string_length` to better clarify what it actually represents (and how it's used by callers). So, there is no buffer overrun risk here.